Collect runtime statistics about phase execution

subpop opened the pull request:
This PR adds the collection of runtime metrics to the "phase runner" (`run.py`). Before each phase, an "event" object is created. The event object contains details about the phase, include start times, exit codes, exception strings, etc. It uses a specialized HTTP client (not `InsightsConnection`) so that statistics can still be reported back to the platform even in the event of a major failure of `InsightsConnection`. It keys off the RHSM server value. If it's not subscription.rhsm.redhat.com, the base URL, authentication scheme and API path prefix are configured to the "Satellite" values. If the RHSM server value is subscription.rhsm.redhat.com, then it examines the insights-client.conf authmethod value, setting up the base URL, authentication, and API path prefix values for either certificate or basic authentication.

**Note**: ~The `default` test will fail until #120 is merged; the `private-rhel-8.3.0` dist-git branch we use to build the RPM from is already updated to expect files that *are* present in #120, but that are absent in this PR. Merging #120 onto master, and then rebasing this PR on top of master will make that pipeline start passing.~

subpop commented:
test this please

gravitypriest commented on src/insights_client/metrics.py:
```diff
@@ -0,0 +1,53 @@
+
+        :param event: a dictionary describing an event object
+        """
+        if self.base_url == "cert-api.access.redhat.com":
```
We'll need to set `self.verify = /etc/insights-client/<whatever that pem file is>` here if connecting to the classic API.

lphiri commented:
@subpop  this PR piqued my interest because of some things I was looking at last week :)

Will this work when an http proxy is required? 
Also, we do have scenarios where the certs just wont work and basic auth must be used (e.g.  cloud instances in AWS etc getting updates from RHUI)

subpop commented:
> @subpop @gravitypriest this PR piqued my interest because of some things I was looking at last week :)
> 
> Will this work when an http proxy is required?

It should. It uses a `requests.Session` underneath, and I *believe* `requests.Session` respects the `HTTPS_PROXY` environment variable.

> Also, we do have scenarios where the certs just wont work and basic auth must be used (e.g. cloud instances in AWS etc getting updates from RHUI)

I'm hoping to avoid any dependency on code inside the egg, in order to make the event tracking isolated from the egg code. I could parse and load the config file to get the authentication mechanism. That way it could be made to work with both certificate and basic authentication.

lphiri commented:
@subpop  - I remember vaguely that we had some really convulated proxy config logic in auto_config  and connection at some point ( we don't just depend on ENV variables but also parse config). If that's been simplified then its good otherwise your specialized connection class may need to do something similar.

gravitypriest commented:
@lphiri I think it's mostly the same, but this may be a good opportunity to revisit that

subpop commented:
I rewrote the URL detection logic, with Satellite in mind. I removed the socket logic that attempts to connect to a list of descending servers. The approach I took varied slightly from the approach in the `auto_config.py` module. It is deliberately simple. It keys off the RHSM server value. If it's *not* `subscription.rhsm.redhat.com`, the base URL, authentication scheme and API path prefix are configured to the "Satellite" values. If the RHSM server value is `subscription.rhsm.redhat.com`, then it examines the `insights-client.conf` `authmethod` value, setting up the base URL, authentication, and API path prefix values for either certificate or basic authentication.

subpop commented:
test this please

subpop commented:
test this please

subpop commented:
@lphiri I rewrote the base URL detection logic. Does this seem more reasonable to you?

subpop commented:
test this please

patchkez commented:
test this please

patchkez commented:
test this please

subpop commented:
test this please

patchkez commented:
test this please

patchkez commented:
test this please

subpop commented:
test this please

subpop commented:
test this please

subpop commented:
test this please

subpop commented:
test this please

subpop commented:
test this please

subpop commented:
test this please
